Added support for PHP 8.1, updated several dependencies

mimmi20 opened the pull request:
Fixes #334

mimmi20 committed:
Added support for PHP 8.1 and updated several dependencies

parent c4c907e092a7f4b61fa976ceacbd705acff1829f
author Thomas Müller <thomas.mueller@jungdms.de> 1638516208 +0100
committer Thomas Müller <thomas.mueller@jungdms.de> 1638986305 +0100

update dependencies

asgrim commented on composer.json:
```diff
@@ -25,34 +25,33 @@
+        "phpstan/phpstan-deprecation-rules": "^1.0.0",
+        "phpstan/phpstan-phpunit": "^1.0.0",
+        "phpunit/phpunit": "^9.5.10",
+        "roave/doctrine-simplecache": "^2.4.0"
```
We should probably depend on something not deprecated any more, I found [PHP-Cache](http://www.php-cache.com/en/latest/) to be the simplest drop-in replacement for `doctrine/cache`

asgrim commented on .github/workflows/continuous-integration.yml:
```diff
@@ -0,0 +1,408 @@
+name: "Continuous Integration"
```
Since this is a library we're interested in maintaining a BC API on, lets add this step too please?

```
  roave-backwards-compatibility-check:
    name: "Check for Backward Compatibility breaks"
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
        with:
          fetch-depth: 0
      - name: "Install PHP"
        uses: shivammathur/setup-php@v2
        with:
          coverage: "none"
          php-version: "7.4" # https://github.com/Roave/BackwardCompatibilityCheck/issues/283
      - name: "Require Roave/BackwardCompatibilityCheck"
        run: "composer require --no-update --no-interaction --prefer-dist --prefer-stable --dev roave/backward-compatibility-check:6.0.x-dev"
      - name: "Composer update with new requirements"
        run: "composer update --no-interaction --prefer-dist --prefer-stable"
      - name: "Check for BC breaks"
        run: "vendor/bin/roave-backward-compatibility-check --format=markdown"
```

If it fails without dev dependencies, you can add the `--install-development-dependencies` to `vendor/bin/roave-backward-compatibility-check`

asgrim commented on composer.json:
```diff
@@ -25,35 +25,35 @@
+        "league/flysystem": "Filesystem cache, Required to use the console commands",
+        "matthiasmullie/scrapbook": "Caching Lib, Required to use the console commands",
+        "monolog/monolog": "Logging Lib, Required to use the console commands",
+        "symfony/console": "Required to use the console commands"
```
I would say `symfony/console` is actually a required dependency, rather than a `suggest`

mimmi20 replied:
I you dont want to use the console commands, then you dont need it.

asgrim replied:
It's difficult to analyse code where dependencies are optional; the CLI is a large part of how you use it IMO, and few would object to symfony/console IMO, even if they didn't already have it heh

asgrim commented on composer.json:
```diff
@@ -24,39 +23,38 @@
+    },
     "require": {
-        "php": ">=7.3.0,<8.1.0",
+        "php": ">=7.4.3,<8.1.0",
```
We should probably widen support to be `>=7.4.3,<8.2.0` now that 8.1 is out

asgrim reviewed (commented):
Mostly looking good here :+1: Could probably do with squashing a few of the commits though to show clearer commit messages :sweat_smile:

mimmi20 commented:
> Mostly looking good here 👍 Could probably do with squashing a few of the commits though to show clearer commit messages

When I tried to force push the rebased branch on the console, I got this error:

```shell
! [remote rejected] updates -> updates (refusing to allow a Personal Access Token to create or update workflow `.github/workflows/continuous-integration.yml` without `workflow` scope)
error: failed to push some refs to 'https://github.com/browscap/browscap-php.git'
```

Normally I use the Github Desktop to push my commits and normal pushes have no issue.

mimmi20 commented:
> Mostly looking good here 👍 Could probably do with squashing a few of the commits though to show clearer commit messages

In my other projects here, I squash the commits here

asgrim committed:
Roave/BCC checker now requires PHP 8+

mimmi20 commented on .github/workflows/continuous-integration.yml:
```diff
@@ -487,7 +487,7 @@ jobs:
 
       - name: "Require Roave/BackwardCompatibilityCheck"
-        run: "composer require --no-update --no-interaction --prefer-dist --prefer-stable --dev \"roave/backward-compatibility-check:6.0.x-dev\""
+        run: "composer require --no-update --no-interaction --prefer-dist --prefer-stable --dev \"roave/backward-compatibility-check:6.0.x-dev\" \"roave/better-reflection:5.0.x-dev\""
```
Do you know when the stable release for these tools is planned?

asgrim replied:
We're trying to get it done ASAP; depending on it in "dev" state is not a big deal, it's just it's a moving target, plus there's a bug I discovered in Roave/BackwardCompatibilityCheck#360

asgrim replied:
@mimmi20 I'll get this fixed upstream ( Roave/BackwardCompatibilityCheck#361 ) and manage the merge of this from now on

asgrim committed:
Remove composer outdated check

It does not make sense to have this check on a library, as the "lowest" dependencies
will ALWAYS be outdated.

asgrim committed:
Added @no-named-arguments annotation to all interfaces

asgrim committed:
Listed BC breaks in 6.0.x

asgrim commented:
BC breaks are expected here, and documented now in `README.md`. Merging into `6.0.x`.

asgrim commented:
@mimmi20 thank you for working on this, it is now merged and released for `6.0.0` :+1:
